core/src/main/java/org/elasticsearch/index/similarity/SimilarityService.java

```diff
@@ -63,6 +63,8 @@ public SimilarityService(IndexSettings indexSettings, Map<String, BiFunction<Str
         Map<String, Settings> similaritySettings = this.indexSettings.getSettings().getGroups(IndexModule.SIMILARITY_SETTINGS_PREFIX);
         for (Map.Entry<String, Settings> entry : similaritySettings.entrySet()) {
             String name = entry.getKey();
+            if(BUILT_IN.containsKey(name))
```

Comment:
@s1monw makes sense to give users a chance to fix this. So then what we are saying here is that we _should_ allow built-in similarities to be overwritten for pre v3 indices? So `SimilarityService.addSimilarities()` may still need to skip overwriting a user defined provider (see https://github.com/elastic/elasticsearch/commit/af0e07a1371de762901a3f6064024e8ef08bb5d6) but only for pre v3 indices. E.g.

```java
if(providers.containsKey(name) && indexVersion.before(Version.V_3_0_0)) {
   continue;
}
``` 

As for moving the validation into the settings infrastructure, it seems `IndexScopedSettings` might be the place but I noticed your comment [about validating analyzers/similarity](https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/common/settings/IndexScopedSettings.java#L136) in there. Wouldn't something like a `SimilarityService.INDEX_SIMILARITY_DEFAULT` setting still be possible?